pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go

```diff
@@ -421,20 +437,37 @@ func GatewayToInstanceMapper(l logr.Logger, client kube_client.Client) kube_hand
 
 		var req []kube_reconcile.Request
 		for _, serviceName := range serviceNames {
+			// TODO: remove in version 2.9.x
 			instances := &mesh_k8s.MeshGatewayInstanceList{}
-			if err := client.List(
-				ctx, instances, kube_client.MatchingFields{serviceKey: serviceName},
-			); err != nil {
+			err := client.List(ctx, instances, kube_client.MatchingFields{serviceKey: serviceName})
+			if err != nil && !kube_apierrs.IsNotFound(err) {
 				l.Error(err, "failed to fetch GatewayInstances")
 			}
+			if len(instances.Items) > 0 {
+				for _, instance := range instances.Items {
+					req = append(req, kube_reconcile.Request{
+						NamespacedName: kube_types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name},
+					})
+				}
+				continue
+			}
 
-			for _, instance := range instances.Items {
-				req = append(req, kube_reconcile.Request{
-					NamespacedName: kube_types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name},
-				})
+			// map generated kuma.io/service
+			name, err := k8s_util.NamespacesNameFromServiceTag(serviceName)
+			if err != nil {
+				l.Error(err, "failed to fetch GatewayInstances")
+				continue
+			}
+			instance := &mesh_k8s.MeshGatewayInstance{}
+			err = client.Get(ctx, name, instance)
+			if err != nil {
+				l.Error(err, "failed to fetch GatewayInstances")
+				continue
 			}
```

Comment:
I think the right behaviour is to enqueue the request and then the reconcile ignores the request if the entity is not found.